python/ql/src/Security/CWE-327/Ssl.qll

```diff
@@ -0,0 +1,221 @@
+      notFlag.getOp() instanceof Invert and
+      notFlag.getOperand() = flag and
+      flag = API::moduleImport("ssl").getMember("OP_NO_" + restriction).getAUse().asExpr() and
+      (
+        aa.getValue() = notFlag
+        or
+        impliesBitSet(aa.getValue(), notFlag, true, true)
+      )
+    )
+  }
+
+  override DataFlow::CfgNode getContext() { result = this }
+
+  override ProtocolVersion getUnrestriction() { result = restriction }
+}
+
+/**
+ * Holds if
+ *   for every bit, _b_:
+ *     `wholeHasBitSet` represents that _b_ is set in `whole`
+ *     implies
+ *     `partHasBitSet` represents that _b_ is set in `part`
+ *
+ * As an example take `whole` = `part1 & part2`. Then
+ * `impliesBitSet(whole, part1, true, true)` holds
+ * because for any bit in `whole`, if that bit is set it must also be set in `part1`.
+ *
+ * Similarly for `whole` = `part1 | part2`. Here
+ * `impliesBitSet(whole, part1, false, false)` holds
+ * because for any bit in `whole`, if that bit is not set, it cannot be set in `part1`.
+ */
+predicate impliesBitSet(BinaryExpr whole, Expr part, boolean partHasBitSet, boolean wholeHasBitSet) {
+  whole.getOp() instanceof BitAnd and
+  (
+    wholeHasBitSet = true and partHasBitSet = true and part in [whole.getLeft(), whole.getRight()]
+    or
+    wholeHasBitSet = true and
+    impliesBitSet([whole.getLeft(), whole.getRight()], part, partHasBitSet, wholeHasBitSet)
+  )
+  or
+  whole.getOp() instanceof BitOr and
+  (
+    wholeHasBitSet = false and partHasBitSet = false and part in [whole.getLeft(), whole.getRight()]
+    or
+    wholeHasBitSet = false and
+    impliesBitSet([whole.getLeft(), whole.getRight()], part, partHasBitSet, wholeHasBitSet)
+  )
+}
+
+class ContextSetVersion extends ProtocolRestriction, ProtocolUnrestriction {
+  ProtocolVersion restriction;
+
+  ContextSetVersion() {
+    exists(DataFlow::AttrWrite aw |
+      aw.getObject().asCfgNode() = node and
+      aw.getAttributeName() = "minimum_version" and
+      aw.getValue() =
+        API::moduleImport("ssl").getMember("TLSVersion").getMember(restriction).getAUse()
+    )
+  }
+
+  override DataFlow::CfgNode getContext() { result = this }
+
+  override ProtocolVersion getRestriction() { result.lessThan(restriction) }
+
+  override ProtocolVersion getUnrestriction() {
+    restriction = result or restriction.lessThan(result)
+  }
+}
+
+class UnspecificSSLContextCreation extends SSLContextCreation, UnspecificContextCreation {
+  UnspecificSSLContextCreation() { library = "ssl" }
+
+  override ProtocolVersion getUnrestriction() {
+    result = UnspecificContextCreation.super.getUnrestriction() and
+    // These are turned off by default
+    // see https://docs.python.org/3/library/ssl.html#ssl-contexts
+    not result in ["SSLv2", "SSLv3"]
+    or
+    // The default argument is TLS and the SSL versions are turned off by default.
+    not exists(this.getProtocol()) and
+    result in ["TLSv1", "TLSv1_1", "TLSv1_2", "TLSv1_3"]
```

Comment:
I'm confused about this part. It seems like both parts of the disjunction deals with the default behavior when the protocol is not specified. And both parts wants to enable everything except `SSLv2`, `SSLv3`, following what (at least) Python 3.9 does.

Did any tests improve from the change in https://github.com/github/codeql/pull/5284/commits/16902c2f56993f4ca29f3f05ca1ce1ac35b656be ?

Reply:
The first disjunct, `UnspecificContextCreation`, is the case where an argument is given, but the argument is a protocol family: either "TLS" or "SSLv23". Here `super.getUnrestriction` relies on `getProtocol`, which requires an argument to be found.

The second disjunct is the case for the default behaviour with no argument given.

This part would likely become clearer, if we were to follow your direction for handling default arguments.

Reply:
I'll have to look this over again when reviewing the full code

Reply:
I have refactored default arguments, so this becomes what you would expect (thanks for calling this out).